### PR TITLE
[EH] Add translator from old to new EH instructions

### PR DESCRIPTION
**Pull request by aheejin**
This translates the old Phase 3 EH instructions, which include `try`, `catch`, `catch_all`, `delegate`, and `rethrow`, into the new EH instructions, which include `try_table` (with `catch` / `catch_ref` / `catch_all` / `catch_all_ref`) and `throw_ref`, passed at the Oct 2023 CG meeting.

This translator can be used as a standalone tool by users of the previous EH toolchain to generate binaries for the new spec without recompiling, and also can be used at the end of the Binaryen pipeline to produce binaries for the new spec while the end-to-end toolchain implementation for the new spec is in progress.

While the goal of this pass is not optimization, this tries to a little better than the most naive implementation, namely by omitting a few instructions where possible and trying to minimize the number of additional locals, because this can be used as a standalone translator or the last stage of the pipeline while we can't post-optimize the results because the whole pipeline (-On) is not ready for the new EH.

### CODE REVIEW

File: src/wasm-builder.h [hunk 1]
```diff
@@ -1328,8 +1328,8 @@ class Builder {
     block->name = name;
     if (append) {
       block->list.push_back(append);
-      block->finalize(type);
     }
+    block->finalize(type);
```
Comment (aheejin) [inline src/wasm-builder.h:1332]: Drive-by fix. We should finalize even if `append` is nullptr

Comment (tlively) [reply src/wasm-builder.h:1332]: Is that true? If we aren't appending anything, then should the block already exist and be finalized or be brand new and finalized by `makeBlock`?

File: test/lit/passes/translate-eh-old-to-new.wast [hunk 2]
```diff
@@ -0,0 +1,1421 @@
...
+
+  ;; CHECK:      (func $delegate-target-outer-try-delegate-concrete (type $2) (result i32)
+  ;; CHECK-NEXT:  (throw_ref
+  ;; CHECK-NEXT:   (block $__binaryen_delegate_caller_target0 (result exnref)
+  ;; CHECK-NEXT:    (return
+  ;; CHECK-NEXT:     (block $outer2 (result i32)
+  ;; CHECK-NEXT:      (throw_ref
+  ;; CHECK-NEXT:       (block $l01 (result exnref)
+  ;; CHECK-NEXT:        (br $outer2
+  ;; CHECK-NEXT:         (try_table (result i32) (catch_all_ref $l01)
```
Comment (kripken) [inline test/lit/passes/translate-eh-old-to-new.wast:1395]: I'm surprised to see a single `try_table` for a testcase that has two `try`s. Is that an optimization or must it be like this for some reason?

File: src/wasm-builder.h (hunk 1 above)
Comment (kripken) [reply src/wasm-builder.h:1332]: I think the old logic here makes sense in that it assumes that if there is a previous block then that block is valid (i.e. finalized), but the comment for the function was not that clear (probably my fault).

@aheejin is this change needed for the new pass for some reason? If so perhaps we can add a new function or rethink this one.

File: test/lit/passes/translate-eh-old-to-new.wast (hunk 2 above)
Comment (aheejin) [reply test/lit/passes/translate-eh-old-to-new.wast]: No it was a bug. Pushed a fix, thanks.

**Review by tlively — COMMENTED** [review]
Haven't taken a look at the tests yet, but here are comments on the code.

File: test/lit/passes/translate-eh-old-to-new.wast [hunk 3]
```diff
@@ -0,0 +1,1425 @@
...
+      )
+      (delegate 0)
+    )
+  )
+
+  ;; CHECK:      (func $delegate-nested-more (type $1)
+  ;; CHECK-NEXT:  (block $outer3
+  ;; CHECK-NEXT:   (block $catch_all4
+  ;; CHECK-NEXT:    (try_table (catch_all $catch_all4)
+  ;; CHECK-NEXT:     (throw_ref
```
Comment (kripken) [inline test/lit/passes/translate-eh-old-to-new.wast:1316]: Reading testcases like this, I get the impression that the new code is less efficient than the old. Is that right? Specifically it looks like to implement a delegate we do a branch and a throw, and before we didn't have this manual throw operation, so the VM could have just jumped along the delegate path.

If that makes sense, a followup question, would new compiler output be able to avoid the extra overhead?

File: src/wasm-builder.h (hunk 1 above)
Comment (aheejin) [reply src/wasm-builder.h:1332]: I think you're right, but things are a little tricky.

So first off, I think we should've passed `name` to `makeBlock` in the first place, because `makeBlock` uses `BranchSeeker` when finalizing only when the block has a name, and we need that `BranchSeeker` to figure out the inner `try_table (catch_ref $label)` is targeting `$label` and its return type is `exnref`.
https://github.com/WebAssembly/binaryen/blob/185019919b1fa2efe8d02056827a31f4f50dd01e/src/wasm/wasm.cpp#L197-L209

But there's another problem. If the passed expression is already a block, `blockifyWithName` reuses it, in which case the passed `type` parameter (in case it is not null) doesn't get used. This doesn't mean the existing block is invalid. For example, in `processDelegateTarget`, the input to `blockfiyWithName` is
```wast
(block
  (try_table (catch_all_ref $label)
    ...
  )
)
```
The `block` here is the body of the outer `try` (which is targeted by the inner `delegate`, which has been converted to `try_table`). The naive output we would want is this:
```wast
(block $label (result exnref)
  (block
    (try_table (catch_all_ref $label)
      ...
    )
  )
)
```
`(block $label)`'s result type is `exnref` because it is targeted by `catch_all_ref`. But with `blockifyWithName`, we can reuse the input block if it is already a block. So we can do
```wast
(block $label (result exnref)
  (try_table (catch_all_ref $label)
    ...
  )
)
```
But in this case we need a refinalization of the `block` with the given type `type`, even if the given block is not invalid, because we are reusing it.

I updated `blockifyWithName` so that it satisfies this, PTAL. So now we pass `name` to `makeBlock`, and also do finalization when `type` is explicitly given.

```diff
--- a/src/wasm-builder.h
+++ b/src/wasm-builder.h
@@ -1323,11 +1323,14 @@ public:
       block = any->dynCast<Block>();
     }
     if (!block || block->name.is()) {
-      block = makeBlock(any);
+      block = makeBlock(name, any);
+    } else {
+      block->name = name;
     }
-    block->name = name;
     if (append) {
       block->list.push_back(append);
+    }
+    if (append || type) {
       block->finalize(type);
     }
     return block;
```

File: src/passes/passes.h [hunk 4]
```diff
@@ -164,6 +164,8 @@ Pass* createStripEHPass();
 Pass* createStubUnsupportedJSOpsPass();
 Pass* createSSAifyPass();
 Pass* createSSAifyNoMergePass();
+Pass* createTranslateEHOldToNewPass();
+Pass* createTranslateEHNewToOldPass();
```
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/passes.h:168]: The second was a TODO plan; see the comments at the top of `TranslateEH.cpp`. But I will delete this for now given that this has not been implemented anyway.

File: src/passes/TranslateEH.cpp [hunk 5]
```diff
@@ -0,0 +1,823 @@
...
+  : public WalkerPass<PostWalker<TranslateEHOldToNew>> {
+  bool isFunctionParallel() override { return true; }
+
+  // Scans and records which try labels are targetted by delegates and rethrows.
+  struct TargetTryLabelScanner : public PostWalker<TargetTryLabelScanner> {
+    TargetTryLabelScanner(Function* func) { walkFunction(func); }
+
+    std::unordered_set<Name> delegateTargets;
+    std::unordered_set<Name> rethrowTargets;
+    bool isTargettedByDelegates(Try* curr) const {
```
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/TranslateEH.cpp:58]: Oh really... Didn't know that. Will fix.

File: src/passes/TranslateEH.cpp [hunk 6]
```diff
@@ -0,0 +1,823 @@
...
+  // time. Because these locals are written and read right after that, we can
+  // reuse these for all extracted values and (extracted values, exnref) tuples
+  // as long as the type matches.
+  Index getScratchLocal(Type type) {
+    auto it = typeToScratchLocal.find(type);
+    if (it != typeToScratchLocal.end()) {
+      return it->second;
+    }
+    typeToScratchLocal[type] = Builder::addVar(getFunction(), type);
+    return typeToScratchLocal[type];
```
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/TranslateEH.cpp:222]: Oh this is neat. Thanks.

File: src/passes/TranslateEH.cpp [hunk 7]
```diff
@@ -0,0 +1,823 @@
...
+        }
+        rethrowTargetToExnrefLocal[curr->name] =
+          exnrefLocals[rethrowTryDepth - 1];
+      }
+    }
+  };
+
+  std::unique_ptr<LabelUtils::LabelManager> labels;
+  std::unique_ptr<TargetTryLabelScanner> labelScanner;
+  std::unique_ptr<ExnrefLocalAssigner> localAssigner;
```
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/TranslateEH.cpp:198]: I used `unique_ptr` mainly to avoid manually deleting those instances. Does `std::optional` delete the pointers for me? I don't think so... Would there be a better way to do this?

[the remark this answers is not in the document]
Comment (tlively) [reply src/passes/TranslateEH.cpp:198]: I mean you could use e.g. `std::optional<LabelUtils::LabelManager> labels;`. It would still manage the lifetime of the object for you automatically and let you initialize it whenever you want, it just wouldn't require a heap allocation.

File: test/lit/passes/translate-eh-old-to-new.wast (hunk 3 above)
Comment (aheejin) [reply test/lit/passes/translate-eh-old-to-new.wast:1316]: Yeah I also (was aware but realize again) that the old spec is very concise.

I ran the translator on an Adobe binary hanging around somewhere in my local machine and the code size increase is 4.2%. We made the switch knowing this will make the code size worse. But one thing I realized was my estimation was incorrect. As you may know, I once estimated the code size increase (for Adobe) to be around 2.1~2.2%, give that we switch to the option that is similar to what we chose and we remove `delegate`.

The most important case, and actually the only case that matters, a simple `try`-`catch_all`-`rethrow`, which is the most common because it is used for destructors. The generated code for other cases are a lot worse due to tuples and such but they are a lot less frequent so they don't affect the code size much.
```wast
try
  ...
catch_all
  ... ;; cleanup
end
```

My (incorrect) estimation (in the doc) was
```wast
block $label0
  try_table (catch_all_ref $label0)
    ...
  end
end
local.set $exn
... ;; cleanup
local.get $exn
throw_ref
```

But this is missing some instructions. The correct version should be
```wast
block $outer
  block $label0
    try_table (catch_all_ref $label0)
      ...
    end
    br $outer
  end
  local.set $exn
  ... ;; cleanup
  local.get $exn
  throw_ref
end
```
So I missed that we need an outer `block` and `end` and a `br` to it in case exceptions are not thrown. Anyway this seems to make the code size increase double. `delegate` transformation seems to be as I expected it to be, so I don't think that estimation was the culprit...

Anyway, this number might go down after optimizations, but at this point, it's worse than I estimated. Not sure if there is anyway we can mitigate this... If we gzip the binaries the size increase goes down to around 2%.

Comment (kripken) [reply test/lit/passes/translate-eh-old-to-new.wast:1316]: Interesting, thanks for the details!

Personally I feel that 2% or so worse (than the 2% we expected) is not terrible since we are getting advantages in return for this change.

However, the delegate stuff worries me, because my naive expectation is that VMs might emit slower code. Specifically delegates seem like they statically told the VM where to go - "if this try catches it, just go to this other try" - but now we need an extra rethrow - "if this try catches it, rethrow right before the try whose catch I want to enter". The VM would need to pattern-match a rethrow that goes straight into a catch to make that efficient, I guess?

Comment (aheejin) [reply test/lit/passes/translate-eh-old-to-new.wast:1316]: I think that's a possibility, but exception code is executed in rare circumstances in the first place, so I wonder if they consist of a significant portion of the runtime. (Desturctors too; `try`-`catch_all` destructors run only when an exception is thrown elsewhere and we need to run destructors before we rethrow the exception to the enclosing scope)

cc @thibaudmichaud ... Would `throw_ref` be slower than `try`-`delegate`, and if so, how much?

Comment (kripken) [reply test/lit/passes/translate-eh-old-to-new.wast:1316]: If this is fast enough in the VM then sorry for the noise here, btw. My understanding of the work the VM does for exceptions is very limited :smile: 

Re-reading the output of this test here (starting on line 1313) I wonder if the delegate which goes to `$l00`, and then throws to be caught immediately again, could avoid the second throw by going directly to `$catch_all4`? We do know this statically so it seems possible. OTOH there is a mismatch because one is a `catch_all` and the other a `catch_all_ref`, but even so, can we jump to a place that does another jump rather than a throw? (Again, all of this is based on my intuition that jumps are faster than throws - if that's inaccurate, feel free to ignore...)

Comment (thibaudmichaud) [reply test/lit/passes/translate-eh-old-to-new.wast:1316]: That is a good point. `delegate` was indeed just a branch, whereas `throw_ref` calls a runtime function which sets up the exception, triggers the unwinder, etc. which is much slower.
I think we could easily turn this runtime call into a branch in the way that @kripken suggested, i.e. by checking if there is an enclosing try scope in the current function and branching to its handler directly.
That would still be two branches instead of one, but that seems harder to optimise for a much smaller benefit.

File: src/wasm-builder.h (hunk 1 above)
Comment (aheejin) [reply src/wasm-builder.h]: Done in #6223, even though I'm not sure how to add a standalone test there... (Will rebase this PR onto that after that lands, because I'm not using any fancy tools supporting stacked PRs)

File: src/passes/TranslateEH.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/TranslateEH.cpp:198]: Ah, `std::optional` also deletes the pointer when it is deleted. But in this case they are not strictly optional; they are just initialized later. Given that `std::optional` also deletes the pointer, I will change the code to use it, but what's a good practice, or criteria, on how to use `std::unique_ptr` vs. `std::optional`?

[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/TranslateEH.cpp]: https://github.com/WebAssembly/binaryen/pull/6210/commits/87f65d34d07282351868d39c7fff9390b7beac00

File: src/passes/TranslateEH.cpp [hunk 8]
```diff
@@ -0,0 +1,823 @@
...
+    ExnrefLocalAssigner(Function* func, TargetTryLabelScanner* labelScanner)
+      : labelScanner(labelScanner) {
+      walkFunction(func);
+    }
+
+    bool hasExnrefLocal(Name rethrowTarget) const {
+      return rethrowTargetToExnrefLocal.find(rethrowTarget) !=
+             rethrowTargetToExnrefLocal.end();
+    }
+    Index getExnrefLocal(Name rethrowTarget) const {
```
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/TranslateEH.cpp]: Done: https://github.com/WebAssembly/binaryen/pull/6210/commits/04606560a563ecbcd9d9603a65e7ed9cb2b34df5

File: src/passes/TranslateEH.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/TranslateEH.cpp]: Done: https://github.com/WebAssembly/binaryen/pull/6210/commits/c44a66d79d7776d719b278b065b8d0b18ee52515

File: test/lit/passes/translate-eh-old-to-new.wast (hunk 2 above)
Comment (aheejin) [reply test/lit/passes/translate-eh-old-to-new.wast]: https://github.com/WebAssembly/binaryen/pull/6210/commits/9eb535bafef7c5636b7ebead3b85fc128d27a5fb

File: src/passes/TranslateEH.cpp [hunk 9]
```diff
@@ -0,0 +1,823 @@
...
+    // (try
+    //   (do
+    //     ...
+    //   )
+    //   (delegate $delegate_target)
+    // )
+    //
+    // to =>
+    //
+    // (try_table (catch_ref $delegate_br_target)
```
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/TranslateEH.cpp:328]: Technically yes in the actual code, because we are getting a unique label every time anyway so that if labels overlap they be affixed by a number. But I'm worried combining them in this example in the comment makes things confusing, because in the 'before' example, we actually have two different labels at that point, one being the old outer `try` (which has yet to be converted to `try_table`'s delegate target label, and the other being the inner `try_table`'s branch target:
```cpp
    // (try $delegate_target (result sometype)
    //   (do
    //     ...
    //     ;; This had originally been an inner try~delegate and has been
    //     ;; already translated to try_table at this point. See
    //     ;; processDelegate() for how it is done.
    //     (try_table (catch_all_ref $delegate_br_target)
    //       ...
    //     )
    //     ...
    //   (catch
    //     ...
    //   )
```

These two are different labels and we have a map to convert one to the other (`delegateTargetToBrTarget`), so I'n worried if using the same labels in the example (and thus in the code) will make things confusing.

File: src/passes/TranslateEH.cpp [hunk 10]
```diff
@@ -0,0 +1,823 @@
...
+    auto shouldBeRef = [&](Expression* catchBody) -> bool {
+      // If this try is targetted by rethrows and those rethrows exist in the
+      // current catch body, we need to use catch_ref/catch_all_ref. By this
+      // point, all rethrows in the catch bodies have been already converted to
+      // throw_refs, so we check the local numbers to see if those (original)
+      // rethrows used to target the current try label.
+      if (localAssigner->hasExnrefLocal(curr->name)) {
+        Index local = localAssigner->getExnrefLocal(curr->name);
+        for (auto* throwRef : FindAll<ThrowRef>(catchBody).list) {
+          if (auto* localGet = throwRef->exnref->cast<LocalGet>()) {
```
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/TranslateEH.cpp:369]: All `ThrowRef` I created in this pass take a `LocalGet` as an argument. See `visitRethrow`. Added a comment: https://github.com/WebAssembly/binaryen/pull/6210/commits/3412caa8a9337a8f9a5335279925b3883df1d071

File: src/passes/TranslateEH.cpp [hunk 11]
```diff
@@ -0,0 +1,823 @@
...
+    //         ...
+    //       )
+    //       ...
+    //       (return)
+    //     )
+    //   )
+    // )
+    //
+    // If sometype (func's type) is none:
+    // INNER_BODY =
```
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/TranslateEH.cpp:763]: Thanks, removed.

File: src/passes/TranslateEH.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (tlively) [reply src/passes/TranslateEH.cpp]: I would say that std::optional is always better unless there’s a specific reason you need to do a heap allocation, e.g. very large data that would overflow the stack if stored in a std::optional or recursive data that requires an indirection to avoid having infinite size.

File: src/passes/TranslateEH.cpp [hunk 12]
```diff
@@ -0,0 +1,823 @@
...
+    //     (block $catch (result tagtype)
+    //       (try_table (catch $e-i32 $catch)
+    //         ...
+    //       )
+    //       (br $outer)
+    //     )
+    //   )
+    //   (use_inst
+    //     (local.get $scratch-i32)
+    //   )
```
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/TranslateEH.cpp:491]: In case of `catch`, we can, because `pop` is always the first instruction in the catch body in the execution order. So if there's a single `catch`, it can be like
```wast
(block $outer
  (use_inst
    (block $l0 (result i32)
      (try_table (catch $l0 $tag)
        ...
      )
      (br $outer)
    )
  )
)
```

But in case of `catch_all` (used for destructors in C++), it's hard, because there would be another instruction (calls to a destructor) before a `throw_ref`, so `throw_ref` is not the first instruction within a `catch_all` body. For example, in case of a single `catch_all`,
```wast
(block $outer
  (call $destructor) ;; We can't do this!
  (throw_ref
    (block $l0 (result exnref)
      (try_table (catch $l0 $tag)
        ...
      )
      (br $outer)
    )
  )
)
```
We can't do this because we can't bypass the `(call $destructor)`.

And really the only thing that matters to the code size for C++ programs is destructor calls with `try`-`catch_all`, because there are so many of them. The number of `try`-`catch`, which is generated from C++ `try`-`catch` clauses, is so small compared to that of `try`-`catch_all`. (J2Wasm currently doesn't even use `catch_all` and a very small number of `catch`es, so we don't need to worry about it)

So if we can't transform `try`-`catch_all`, I think keeping the current transformation is simpler, because we can apply the same way of transformation to all `catch` variants. If we do one thing for `catch` and something else for `catch_all_ref` that will be more complicated and that doesn't decrease code size much anyway. Also after implementing Binaryen support for optimization pipeline, I think it can take care of those `catch`es later.

File: src/passes/TranslateEH.cpp [hunk 13]
```diff
@@ -0,0 +1,823 @@
...
+    //   (local.set $exn
+    //     (block $catch_all (result exnref)
+    //       (try_table (catch_all_ref $catch_all)
+    //         ...
+    //       )
+    //       (br $outer)
+    //     )
+    //   )
+    //   (throw_ref
+    //     (local.get $exn)
```
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/TranslateEH.cpp:520]: See https://github.com/WebAssembly/binaryen/pull/6210#discussion_r1452941645. Please let me know if I'm missing something.

File: test/lit/passes/translate-eh-old-to-new.wast (hunk 3 above)
Comment (aheejin) [reply test/lit/passes/translate-eh-old-to-new.wast:1316]: @kripken I think we can do this in case the targeted `try` has `catch_all`.

Original:
```wast
(try $l0
  (do
    (try
      (do)
      (delegate $l0)
    )
  )
)
```

What we do now:
```wast
(block $outer
  (block $catch_all0
    (try_table (catch_all $catch_all0)
      (throw_ref
        (block $catch_all1 (result exnref)
          (try_table (catch_all_ref $catch_all1)
          )
          (br $outer)
        )
      )
    )
  )
)
```

What you suggested:
```wast
(block $outer
  (block $catch_all0
    (try_table (catch_all $catch_all0)
      (try_table (catch_all_ref $catch_all0)
      )
      (br $outer)
    )
  )
)
```

Did I get what you suggested correctly?

I think this is worth doing not only from the performance perspective, which I worry less about because the exception paths are executed rarely anyway, but for the code size too, given that it can reduce a `throw_ref` and `block`/`end`. But as I said, this is only possible only when the target `try` has `catch_all`. Do you mind if I do it as a follow-up?

Comment (aheejin) [reply test/lit/passes/translate-eh-old-to-new.wast:1316]: And about the "correct code" I wrote in https://github.com/WebAssembly/binaryen/pull/6210#discussion_r1446748017, I think actually we can reduce `local.set $exn` and `local.get $exn` pair if we run `local2stack` optimization in `optimize-stack-ir`, which brings down the additional number of instruction to only (?) one more than my previous (incorrect) estimation. We don't support the new EH in the stack IR yet, but adding its support to the stack IR should be simple. Supporting the new EH in the whole optimization pipeline will take more time though.

So this
```wast
block $outer
  block $label0
    try_table (catch_all_ref $label0)
      ...
    end
    br $outer
  end
  local.set $exn
  ... ;; cleanup
  local.get $exn
  throw_ref
end
```

can be
```wast
block $outer
  block $label0
    try_table (catch_all_ref $label0)
      ...
    end
    br $outer
  end
  ... ;; cleanup
  throw_ref
end
```

Comment (kripken) [reply test/lit/passes/translate-eh-old-to-new.wast:1316]: @aheejin 

> Did I get what you suggested correctly?

Yes, exactly.

> Do you mind if I do it as a follow-up?

Sounds good.

> this is only possible only when the target try has catch_all.

To make sure I follow, that is in the example that has `catch_all`, but if it were a catch of a specific tag, that would need to match between the two trys? (But maybe we can optimize in some cases even without a match?)

File: src/passes/TranslateEH.cpp [hunk 14]
```diff
@@ -0,0 +1,822 @@
+/*
+ * Copyright 2023 WebAssembly Community Group participants
```
Comment (kripken) [inline src/passes/TranslateEH.cpp:2]:
```suggestion
 * Copyright 2024 WebAssembly Community Group participants
```

File: src/passes/TranslateEH.cpp [hunk 15]
```diff
@@ -0,0 +1,822 @@
...
+  // )
+  struct ExnrefLocalAssigner : public PostWalker<ExnrefLocalAssigner> {
+    TargetTryLabelScanner* labelScanner = nullptr;
+
+    std::vector<Index> exnrefLocals;
+    std::unordered_map<Name, Index> rethrowTargetToExnrefLocal;
+
+    // Depth of the current try nest, when only counting trys targeted with
+    // rethrows.
+    size_t rethrowTryDepth = 0;
```
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/TranslateEH.cpp:143]: True, but I'm not sure we can run `coalesce-locals` or `reorder-locals` with the new EH at this point yet... Supporting it in them wouldn't be a ton of work though. I wanted to make the translator emit maybe not the best but ok code at this point without spending more time on the optimization pipeline.

Also if we assign a local to each `try` the number of locals can be prohibitive. `CoalesceLocal.cpp` says https://github.com/WebAssembly/binaryen/blob/2dfa01f09071ab78e8134fee5c96bcd022330635/src/passes/CoalesceLocals.cpp#L22-L25, which means we should run `simplify-locals` too? 

Anyway, given that the optimization pipeline is not ready for the new EH yet, and also given that this local assignment in this pass is not very complicated, I am inclined to keep this at least for the moment, even though maybe we can delete that later after the opt pipeline is ready. WDYT?

File: src/passes/TranslateEH.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (aheejin) [reply src/passes/TranslateEH.cpp:369]: Ah, good point. Will remove.

File: test/lit/passes/translate-eh-old-to-new.wast (hunk 3 above)
Comment (aheejin) [reply test/lit/passes/translate-eh-old-to-new.wast:1316]: > To make sure I follow, that is in the example that has `catch_all`, but if it were a catch of a specific tag, that would need to match between the two trys? (But maybe we can optimize in some cases even without a match?)

To be more precise, I think this optimization is possible only when the targeted `try` has a single `catch_all`; `try`-`catch`-`catch_all` wouldn't count, given that we don't know whether this exception is gonna be caught in the `catch` or `catch_all`. We can only branch to somewhere when we are certain the exception will be caught there.

I'm not sure what you mean by a match between two `try`s. The inner one is a `try`-`delegate` here, so it can't have tags I think..?

Comment (aheejin) [reply test/lit/passes/translate-eh-old-to-new.wast:1316]: Also I think we can optimize an inner `try`-`delegate` targeting an outer `try`-`delegate` too... because the destination for the inner `delegate` should be the same as the destination for the outer `delegate`, wherever that is. In case the outer `delegate` targets another `try`-`catch_all`, the both inner and outer `delegate` can be optimized to target the third `catch_all`... which gets complicated 🫠

File: src/passes/TranslateEH.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/TranslateEH.cpp:143]: sgtm, this seems fine for now.

About that comment, in general we'd need `simplify-locals`, but I think the locals added in this pass will not benefit from that particular pass just because of the form they are in. I could be wrong though...

File: test/lit/passes/translate-eh-old-to-new.wast (hunk 3 above)
Comment (kripken) [reply test/lit/passes/translate-eh-old-to-new.wast:1316]: Right, sorry, I always found the delegate stuff confusing... I was confused because I thought for some reason a delegate was for a particular tag or such. Yeah, sounds like we need a catch-all in the outer one then, as you said.

Btw, one possible option for later optimization work here is to do less in this pass and add general `TryTable` optimizations elsewhere. For example `RemoveUnsedBrs` has a jump-threading section which might be adapted to also thread exceptions perhaps (since exception control flow is just branches, it could fit there, I hope). Anyhow, just a thought for later maybe.

Comment (aheejin) [reply test/lit/passes/translate-eh-old-to-new.wast:1316]: I was under the impression that it'd be easier to do things here because this pass has the access to the original `try`-`delegate` structure, which I thought was simpler and didn't involve other instructions such as `block`s. But maybe things are easier to maintain if we use something like jump threading after all..? But even if we use jump threading I guess we should do at least certain things here because jump threading itself wouldn't take care of `throw_ref`s and `catch_all_ref`s.

Comment (kripken) [reply test/lit/passes/translate-eh-old-to-new.wast:1316]: Makes sense, yeah, probably doing some stuff here and some elsewhere is best. This pass should definitely use the delegate-specific info it knows, as you said.

But in general I was thinking that in the long term we will need this pass less, while any general-purpose optimizations on TryTable would be helpful forever. So I'd lean in that direction where possible (but I don't know where that line is).

**Comment by aheejin** [comment]
@tlively Any more concerns on the PR?

**Comment by aheejin** [comment]
I added at last one instruction that can throw `like a `call` within `try` bodies in the tests. It is more realistic, and as we will see later with other optimizations, without any throwable instructions within a `try` body, CFG doesn't show the all possible paths.

**Review by tlively — APPROVED** [review]
No concerns, LGTM!
